File: src/main/java/com/indico/query/TrainingModelWithProgressQuery.java
package com.indico.query;

import com.apollographql.apollo.ApolloCall;
import com.apollographql.apollo.api.Response;
import com.indico.Async;
import com.indico.IndicoClient;
import com.indico.ModelGroupProgressGraphQLQuery;
import com.indico.Query;
import com.indico.entity.Model;
import com.indico.entity.TrainingProgress;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;

public class TrainingModelWithProgressQuery implements Query<Model> {

    private int id;
    private String name;
    private final IndicoClient client;

    public TrainingModelWithProgressQuery(IndicoClient client) {
        this.client = client;
    }

    /**
     * Use to query TrainingModelWithProgress by id
     *
     * @param id
     * @return TrainingModelWithProgressQuery
     */
    public TrainingModelWithProgressQuery id(int id) {
        this.id = id;
        return this;
    }

    /**
     * Use to query TrainingModelWithProgress by name
     *
     * @param name
     * @return TrainingModelWithProgressQuery
     */
    public TrainingModelWithProgressQuery name(String name) {
        this.name = name;
        return this;
    }

    /**
     * Queries the server and returns Training Model
     *
     * @return Model
     */
    @Override
    public Model query() {
        ApolloCall<ModelGroupProgressGraphQLQuery.Data> apolloCall = this.client.apolloClient.query(ModelGroupProgressGraphQLQuery.builder()
                .id(this.id)
                .build());
        Response<ModelGroupProgressGraphQLQuery.Data> response = (Response<ModelGroupProgressGraphQLQuery.Data>) Async.executeSync(apolloCall).join();
        List<ModelGroupProgressGraphQLQuery.ModelGroup> modelGroups = response.data().modelGroups().modelGroups();
        if (modelGroups.size() != 1) {
            throw new RuntimeException("Cannot find Model Group");
        }
        List<ModelGroupProgressGraphQLQuery.Model> models = modelGroups.get(0).models();
        ModelGroupProgressGraphQLQuery.Model model = Collections.max(models, Comparator.comparingInt(m -> m.id()));
        TrainingProgress progress = new TrainingProgress(model.trainingProgress().percentComplete());
        return new Model.Builder()
                .id(model.id())
                .status(model.status())
                .trainingProgress(progress)
                .build();
    }
}

File: src/main/java/com/indico/query/TrainingModelWithProgressQuery.kt
package com.indico.query

import com.expediagroup.graphql.client.jackson.types.OptionalInput
import com.indico.IndicoClient
import com.indico.entity.Model
import com.indico.entity.TrainingProgress
import com.indico.exceptions.IndicoQueryException
import com.indico.graphql.ModelGroupProgressGraphQLQuery
import java.util.*

class TrainingModelWithProgressQuery(private val client: IndicoClient) :
    Query<Model?, ModelGroupProgressGraphQLQuery.Result>() {
    private var id = 0
    private var name: String? = null

    /**
     * Use to query TrainingModelWithProgress by id
     *
     * @param id
     * @return TrainingModelWithProgressQuery
     */
    fun id(id: Int): TrainingModelWithProgressQuery {
        this.id = id
        return this
    }

    /**
     * Use to query TrainingModelWithProgress by name
     *
     * @param name
     * @return TrainingModelWithProgressQuery
     */
    @Deprecated("not supported")
    fun name(name: String?): TrainingModelWithProgressQuery {
        this.name = name
        return this
    }

    /**
     * Queries the server and returns Training Model
     *
     * @return Model
     */
    override fun query(): Model {
        return try {
            val call = ModelGroupProgressGraphQLQuery(ModelGroupProgressGraphQLQuery.Variables(
                id = if(id != null) OptionalInput.Defined(id) else OptionalInput.Undefined
            ))
            val response = client.execute(call)
            handleErrors(response)
            val modelGroups = response.data?.modelGroups?.modelGroups?:
            throw IndicoQueryException("Error fetching model group for model $id")
            if (modelGroups.isEmpty()) {
                throw IndicoQueryException("Cannot find Model Group $id")
            }
            val models = modelGroups[0]?.models?: throw IndicoQueryException("Error fetching models for model group $id")
            val model = models.maxByOrNull { m -> m!!.id!! }

            val percentComplete = model?.trainingProgress?.percentComplete?:
            throw IndicoQueryException("Model progress percent is missing for $id")
            val progress = TrainingProgress(percentComplete.toDouble())
            Model.Builder()
                .id(model.id!!)
                .status(model.status!!)
                .trainingProgress(progress)
                .build()
        }catch (ex: RuntimeException) {
            throw IndicoQueryException("Call to check Model progress failed", ex)
        }
    }

    override fun refresh(obj: Model?): Model? {
        return obj
    }

}